File: FHD/mobile-android/app/src/main/java/com/xiuci/xcagi/mobile/core/speech/VoiceInputSheet.kt
package com.xiuci.xcagi.mobile.core.speech

import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.spring
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Check
import androidx.compose.material.icons.filled.Close
import androidx.compose.material.icons.filled.Mic
import androidx.compose.material.icons.filled.Refresh
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.scale
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import com.xiuci.xcagi.mobile.ui.theme.Elevation
import com.xiuci.xcagi.mobile.ui.theme.Spacing

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun VoiceInputSheet(
    onResult: (String) -> Unit,
    onDismiss: () -> Unit,
) {
    val context = androidx.compose.ui.platform.LocalContext.current
    val helper = remember { SpeechRecognizerHelper(context) }
    val state by helper.state.collectAsState()
    val partial by helper.partialResult.collectAsState()
    val final by helper.finalResult.collectAsState()
    val rms by helper.rms.collectAsState()
    val errorText by helper.errorText.collectAsState()
    val sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true)
    val finalText = final.trim()
    val recognizedText = finalText.ifBlank { partial }
    val hasFinalResult = finalText.isNotBlank()

    LaunchedEffect(Unit) { helper.start() }

    DisposableEffect(Unit) {
        onDispose { helper.destroy() }
    }

    fun dismiss() {
        helper.destroy()
        onDismiss()
    }

    fun primaryAction() {
        when {
            hasFinalResult -> {
                onResult(finalText)
                dismiss()
            }
            state == SpeechState.ERROR -> helper.start()
            state == SpeechState.PROCESSING -> Unit
            else -> helper.stop()
        }
    }

    ModalBottomSheet(
        onDismissRequest = { dismiss() },
        sheetState = sheetState,
        containerColor = VoiceInputDesign.palette(state).sheetBackground,
        tonalElevation = Elevation.level3,
        shape = RoundedCornerShape(
            topStart = VoiceInputDesign.sheetTopCornerRadius,
            topEnd = VoiceInputDesign.sheetTopCornerRadius,
        ),
        dragHandle = { VoiceDragHandle() },
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(horizontal = VoiceInputDesign.sheetHorizontalPadding)
                .padding(top = VoiceInputDesign.sheetTopPadding, bottom = VoiceInputDesign.sheetBottomPadding),
            verticalArrangement = Arrangement.spacedBy(Spacing.lg),
        ) {
            VoiceSheetHeader(
                state = state,
                errorText = errorText,
                hasResult = hasFinalResult,
                onDismiss = { dismiss() },
            )
            VoiceListeningCard(
                state = state,
                rms = rms,
                preview = recognizedText,
                errorText = errorText,
                hasResult = hasFinalResult,
            )
            VoiceActionRow(
                state = state,
                hasResult = hasFinalResult,
                onPrimary = { primaryAction() },
                onCancel = { dismiss() },
            )
        }
    }
}

@Composable
private fun VoiceDragHandle() {
    val palette = VoiceInputDesign.palette(SpeechState.IDLE)
    Box(
        modifier = Modifier
            .padding(top = Spacing.sm, bottom = Spacing.xs)
            .size(width = VoiceInputDesign.dragHandleWidth, height = VoiceInputDesign.dragHandleHeight)
            .clip(RoundedCornerShape(VoiceInputDesign.dragHandleCornerRadius))
            .background(palette.dragHandle),
    )
}

@Composable
private fun VoiceSheetHeader(
    state: SpeechState,
    errorText: String,
    hasResult: Boolean,
    onDismiss: () -> Unit,
) {
    Row(
        modifier = Modifier.fillMaxWidth(),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Column(Modifier.weight(1f)) {
            Text(
                text = "语音输入",
                style = MaterialTheme.typography.titleMedium,
                fontWeight = FontWeight.SemiBold,
                color = MaterialTheme.colorScheme.onSurface,
            )
            Spacer(Modifier.height(Spacing.xs))
            VoiceStatusPill(state = state, errorText = errorText, hasResult = hasResult)
        }
        IconButton(onClick = onDismiss) {
            Icon(
                Icons.Default.Close,
                contentDescription = "关闭",
                tint = MaterialTheme.colorScheme.onSurfaceVariant,
            )
        }
    }
}

@Composable
private fun VoiceStatusPill(state: SpeechState, errorText: String, hasResult: Boolean) {
    val palette = VoiceInputDesign.palette(state)
    Surface(
        shape = RoundedCornerShape(999.dp),
        color = palette.statusBackground,
    ) {
        Text(
            text = VoiceInputDesign.statusLabel(state, errorText, hasResult),
            modifier = Modifier.padding(horizontal = Spacing.md, vertical = 5.dp),
            style = MaterialTheme.typography.labelMedium,
            color = palette.statusForeground,
            maxLines = 1,
            overflow = TextOverflow.Ellipsis,
        )
    }
}

@Composable
private fun VoiceListeningCard(
    state: SpeechState,
    rms: Float,
    preview: String,
    errorText: String,
    hasResult: Boolean,
) {
    val isListening = state == SpeechState.LISTENING
    val palette = VoiceInputDesign.palette(state)
    val micScale by animateFloatAsState(
        targetValue = if (isListening) 1f + rms * 0.22f else 1f,
        animationSpec = spring(),
        label = "voiceMicScale",
    )

    val cardShape = RoundedCornerShape(VoiceInputDesign.cardCornerRadius)
    Box(
        modifier = Modifier
            .fillMaxWidth()
            .clip(cardShape)
            .background(palette.cardBackground)
            .border(1.dp, palette.cardBorder, cardShape),
    ) {
        Column(
            modifier = Modifier.padding(horizontal = Spacing.lg, vertical = VoiceInputDesign.cardVerticalPadding),
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.spacedBy(Spacing.md),
        ) {
            Row(
                modifier = Modifier.fillMaxWidth(),
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(Spacing.lg),
            ) {
                Box(
                    modifier = Modifier.size(VoiceInputDesign.micOuterSize),
                    contentAlignment = Alignment.Center,
                ) {
                    if (isListening) VoicePulse(palette.pulse)
                    Box(
                        modifier = Modifier
                            .size(VoiceInputDesign.micInnerSize)
                            .clip(CircleShape)
                            .background(palette.micBackground),
                        contentAlignment = Alignment.Center,
                    ) {
                        Icon(
                            Icons.Default.Mic,
                            contentDescription = "语音输入",
                            tint = palette.micForeground,
                            modifier = Modifier
                                .size(VoiceInputDesign.micIconSize)
                                .scale(micScale),
                        )
                    }
                }
                VoiceWaveform(
                    level = if (isListening) rms else 0.16f,
                    color = palette.waveform,
                    modifier = Modifier.weight(1f),
                )
            }

            val previewShape = RoundedCornerShape(VoiceInputDesign.previewCornerRadius)
            Box(
                modifier = Modifier
                    .fillMaxWidth()
                    .heightIn(min = VoiceInputDesign.previewMinHeight)
                    .clip(previewShape)
                    .background(palette.previewBackground),
            ) {
                val displayText = preview.ifBlank { VoiceInputDesign.statusLabel(state, errorText, hasResult) }
                Text(
                    text = displayText,
                    modifier = Modifier.padding(horizontal = Spacing.lg, vertical = Spacing.md),
                    style = MaterialTheme.typography.bodyMedium,
                    color = if (preview.isBlank()) palette.previewPlaceholder else palette.previewForeground,
                    textAlign = TextAlign.Start,
                )
            }
        }
    }
}

@Composable
private fun VoicePulse(color: androidx.compose.ui.graphics.Color) {
    val pulse = rememberInfiniteTransition(label = "voicePulse")
    listOf(0, 520).forEach { delayMs ->
        val p by pulse.animateFloat(
            initialValue = 0f,
            targetValue = 1f,
            animationSpec = infiniteRepeatable(
                animation = tween(1320, delayMillis = delayMs),
                repeatMode = RepeatMode.Restart,
            ),
            label = "voicePulse$delayMs",
        )
        Box(
            modifier = Modifier
                .size((VoiceInputDesign.micInnerSize.value + 30f * p).dp)
                .clip(CircleShape)
                .background(color.copy(alpha = (1f - p) * 0.15f)),
        )
    }
}

@Composable
private fun VoiceWaveform(
    level: Float,
    color: androidx.compose.ui.graphics.Color,
    modifier: Modifier = Modifier,
) {
    Row(
        modifier = modifier.height(VoiceInputDesign.waveformHeight),
        horizontalArrangement = Arrangement.spacedBy(VoiceInputDesign.waveformBarGap),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        VoiceInputDesign.waveformWeights.forEachIndexed { index, weight ->
            val targetHeight = 8f + (level.coerceIn(0f, 1f) * 30f * weight)
            val animatedHeight by animateFloatAsState(
                targetValue = targetHeight,
                animationSpec = spring(),
                label = "voiceBar$index",
            )
            Box(
                modifier = Modifier
                    .width(VoiceInputDesign.waveformBarWidth)
                    .height(animatedHeight.dp)
                    .clip(RoundedCornerShape(999.dp))
                    .background(color.copy(alpha = 0.34f + weight * 0.42f)),
            )
        }
    }
}

@Composable
private fun VoiceActionRow(
    state: SpeechState,
    hasResult: Boolean,
    onPrimary: () -> Unit,
    onCancel: () -> Unit,
) {
    val isProcessing = state == SpeechState.PROCESSING
    val palette = VoiceInputDesign.palette(state)
    Row(
        modifier = Modifier.fillMaxWidth(),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(Spacing.md),
    ) {
        TextButton(
            onClick = onCancel,
            modifier = Modifier
                .weight(1f)
                .height(VoiceInputDesign.actionHeight),
        ) {
            Text("取消", color = palette.cancelForeground)
        }
        Button(
            onClick = onPrimary,
            enabled = !isProcessing,
            modifier = Modifier
                .weight(1f)
                .height(VoiceInputDesign.actionHeight),
            colors = ButtonDefaults.buttonColors(
                containerColor = palette.primaryBackground,
                contentColor = palette.primaryForeground,
                disabledContainerColor = palette.disabledPrimaryBackground,
                disabledContentColor = palette.disabledPrimaryForeground,
            ),
            shape = RoundedCornerShape(999.dp),
        ) {
            val icon = when {
                hasResult -> Icons.Default.Check
                state == SpeechState.ERROR -> Icons.Default.Refresh
                else -> Icons.Default.Mic
            }
            Icon(icon, contentDescription = null, modifier = Modifier.size(18.dp))
            Spacer(Modifier.width(Spacing.xs))
            Text(VoiceInputDesign.primaryActionLabel(state, hasResult))
        }
    }
}

File: FHD/mobile-flutter-poc/lib/src/features/voice/voice_input_sheet.dart
import 'package:flutter/material.dart';
import 'package:speech_to_text/speech_recognition_error.dart';
import 'package:speech_to_text/speech_recognition_result.dart';
import 'package:speech_to_text/speech_to_text.dart';

import '../../theme/app_theme.dart';

enum _SpeechUiState { idle, listening, processing, error }

class VoiceInputDesign {
  static const sheetTopCornerRadius = 28.0;
  static const sheetHorizontalPadding = 20.0;
  static const sheetTopPadding = 6.0;
  static const sheetBottomPadding = 28.0;
  static const cardCornerRadius = 22.0;
  static const cardVerticalPadding = 18.0;
  static const micOuterSize = 92.0;
  static const micInnerSize = 64.0;
  static const micIconSize = 30.0;
  static const waveformHeight = 42.0;
  static const waveformBarWidth = 4.0;
  static const waveformBarGap = 5.0;
  static const actionHeight = 48.0;
  static const dragHandleWidth = 42.0;
  static const dragHandleHeight = 5.0;
  static const dragHandleCornerRadius = 999.0;
  static const previewMinHeight = 42.0;
  static const previewCornerRadius = 16.0;
  static const pulseExpansion = 30.0;
  static const pulseMaxAlpha = 0.15;
  static const pulseDurationMs = 1320;
  static const pulseSecondDelayMs = 520;

  static const waveformWeights = <double>[
    0.36,
    0.52,
    0.78,
    1,
    0.72,
    0.9,
    0.6,
    0.42,
  ];

  @visibleForTesting
  static Map<String, double> dimensionTokensForTest() {
    return {
      'sheetTopCornerRadius': sheetTopCornerRadius,
      'sheetHorizontalPadding': sheetHorizontalPadding,
      'sheetTopPadding': sheetTopPadding,
      'sheetBottomPadding': sheetBottomPadding,
      'cardCornerRadius': cardCornerRadius,
      'cardVerticalPadding': cardVerticalPadding,
      'micOuterSize': micOuterSize,
      'micInnerSize': micInnerSize,
      'micIconSize': micIconSize,
      'waveformHeight': waveformHeight,
      'waveformBarWidth': waveformBarWidth,
      'waveformBarGap': waveformBarGap,
      'actionHeight': actionHeight,
      'dragHandleWidth': dragHandleWidth,
      'dragHandleHeight': dragHandleHeight,
      'dragHandleCornerRadius': dragHandleCornerRadius,
      'previewMinHeight': previewMinHeight,
      'previewCornerRadius': previewCornerRadius,
      'pulseExpansion': pulseExpansion,
      'pulseMaxAlpha': pulseMaxAlpha,
      'pulseDurationMs': pulseDurationMs.toDouble(),
      'pulseSecondDelayMs': pulseSecondDelayMs.toDouble(),
    };
  }

  @visibleForTesting
  static List<double> waveformWeightsForTest() {
    return List.unmodifiable(waveformWeights);
  }
}

class VoiceInputSheet extends StatefulWidget {
  const VoiceInputSheet({super.key, required this.onResult});

  final ValueChanged<String> onResult;

  @override
  State<VoiceInputSheet> createState() => _VoiceInputSheetState();
}

class _VoiceInputSheetState extends State<VoiceInputSheet> {
  final _speech = SpeechToText();
  var _state = _SpeechUiState.idle;
  var _partial = '';
  var _final = '';
  var _errorText = '';
  var _soundLevel = 0.16;

  bool get _hasResult => _final.trim().isNotEmpty;
  String get _recognizedText => _final.trim().isNotEmpty ? _final : _partial;

  @override
  void initState() {
    super.initState();
    WidgetsBinding.instance.addPostFrameCallback((_) => _startListening());
  }

  @override
  void dispose() {
    _speech.cancel();
    super.dispose();
  }

  Future<void> _startListening() async {
    setState(() {
      _state = _SpeechUiState.listening;
      _partial = '';
      _final = '';
      _errorText = '';
      _soundLevel = 0.16;
    });
    try {
      final available = await _speech.initialize(
        onStatus: _handleStatus,
        onError: _handleError,
      );
      if (!mounted) return;
      if (!available) {
        setState(() {
          _state = _SpeechUiState.error;
          _errorText = '没听清';
        });
        return;
      }
      await _speech.listen(
        onResult: _handleResult,
        onSoundLevelChange: (level) {
          if (!mounted || _state != _SpeechUiState.listening) return;
          setState(() => _soundLevel = (level / 10).abs().clamp(0.16, 1.0));
        },
        listenOptions: SpeechListenOptions(
          listenMode: ListenMode.dictation,
          partialResults: true,
          cancelOnError: false,
        ),
      );
    } catch (_) {
      if (!mounted) return;
      setState(() {
        _state = _SpeechUiState.error;
        _errorText = '没听清';
      });
    }
  }

  void _handleResult(SpeechRecognitionResult result) {
    if (!mounted) return;
    final words = result.recognizedWords.trim();
    setState(() {
      if (result.finalResult) {
        _final = words;
        _state = words.isEmpty ? _SpeechUiState.error : _SpeechUiState.idle;
        if (words.isEmpty) _errorText = '没听清';
      } else {
        _partial = words;
      }
    });
  }

  void _handleStatus(String status) {
    if (!mounted) return;
    if (status == 'listening') {
      setState(() => _state = _SpeechUiState.listening);
      return;
    }
    if (status == 'done' || status == 'notListening') {
      setState(() {
        if (_hasResult) {
          _state = _SpeechUiState.idle;
        } else if (_partial.trim().isNotEmpty) {
          _final = _partial.trim();
          _state = _SpeechUiState.idle;
        } else if (_state == _SpeechUiState.processing) {
          _state = _SpeechUiState.error;
          _errorText = '没听清';
        }
      });
    }
  }

  void _handleError(SpeechRecognitionError error) {
    if (!mounted) return;
    setState(() {
      _state = _SpeechUiState.error;
      _errorText = _voiceErrorText(error.errorMsg);
    });
  }

  Future<void> _primaryAction() async {
    if (_hasResult) {
      widget.onResult(_final.trim());
      if (mounted) Navigator.of(context).pop();
      return;
    }
    if (_state == _SpeechUiState.error) {
      await _startListening();
      return;
    }
    if (_state == _SpeechUiState.processing) return;
    setState(() => _state = _SpeechUiState.processing);
    try {
      await _speech.stop();
    } catch (_) {
      if (!mounted) return;
      setState(() {
        _state = _SpeechUiState.error;
        _errorText = '没听清';
      });
    }
  }

  void _cancel() {
    _speech.cancel();
    Navigator.of(context).pop();
  }

  @override
  Widget build(BuildContext context) {
    final colors = AppTheme.colors(context);
    final palette = _VoicePalette.of(context, _state);
    return SafeArea(
      top: false,
      child: Padding(
        padding: EdgeInsets.only(
          left: VoiceInputDesign.sheetHorizontalPadding,
          right: VoiceInputDesign.sheetHorizontalPadding,
          top: VoiceInputDesign.sheetTopPadding,
          bottom: MediaQuery.of(context).viewInsets.bottom +
              VoiceInputDesign.sheetBottomPadding,
        ),
        child: Column(
          mainAxisSize: MainAxisSize.min,
          crossAxisAlignment: CrossAxisAlignment.stretch,
          children: [
            Center(
              child: Container(
                width: VoiceInputDesign.dragHandleWidth,
                height: VoiceInputDesign.dragHandleHeight,
                margin: const EdgeInsets.only(bottom: 16),
                decoration: BoxDecoration(
                  color: colors.divider.withValues(alpha: 0.72),
                  borderRadius: BorderRadius.circular(
                    VoiceInputDesign.dragHandleCornerRadius,
                  ),
                ),
              ),
            ),
            Row(
              children: [
                Expanded(
                  child: Column(
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: [
                      Text(
                        '语音输入',
                        style: TextStyle(
                          color: colors.textPrimary,
                          fontSize: 17,
                          height: 1.29,
                          fontWeight: FontWeight.w600,
                          letterSpacing: 0,
                        ),
                      ),
                      const SizedBox(height: 6),
                      _VoiceStatusPill(
                        text: _voiceStatusLabel(_state, _errorText, _hasResult),
                        palette: palette,
                      ),
                    ],
                  ),
                ),
                IconButton(
                  onPressed: _cancel,
                  icon: const Icon(Icons.close),
                  tooltip: '关闭',
                ),
              ],
            ),
            const SizedBox(height: 18),
            _VoiceListeningCard(
              state: _state,
              level: _soundLevel,
              preview: _recognizedText,
              statusText: _voiceStatusLabel(_state, _errorText, _hasResult),
              palette: palette,
              weights: VoiceInputDesign.waveformWeights,
            ),
            const SizedBox(height: 18),
            Row(
              children: [
                Expanded(
                  child: SizedBox(
                    height: VoiceInputDesign.actionHeight,
                    child: TextButton(
                      onPressed: _cancel,
                      child: const Text('取消'),
                    ),
                  ),
                ),
                const SizedBox(width: 12),
                Expanded(
                  child: SizedBox(
                    height: VoiceInputDesign.actionHeight,
                    child: FilledButton.icon(
                      onPressed: _state == _SpeechUiState.processing
                          ? null
                          : _primaryAction,
                      icon: Icon(
                        _hasResult
                            ? Icons.check
                            : _state == _SpeechUiState.error
                                ? Icons.refresh
                                : Icons.mic,
                        size: 18,
                      ),
                      label: Text(_voicePrimaryLabel(_state, _hasResult)),
                    ),
                  ),
                ),
              ],
            ),
          ],
        ),
      ),
    );
  }
}

part 'voice_input_sheet_widgets.part.dart';
